### PR TITLE
Add a new Security interview process

### PR DESCRIPTION
**Pull request by wadells**
This PR contains proposal the candidate facing documentation for two new *synchronous* security engineer hiring processes.

One is application & golang focused.  The other is cloud & terraform focused.

## Notes to the Reviewer

More detail on the design backing these candidate facing documents can be found in [this google doc](https://docs.google.com/document/d/1Ns4nvS5XBxDdje152UteeYUL4PlA85ZLfccxkHER8Kw/edit#).

As we designed this challenge, we aimed for the following goals:

* Under 8 hours of candidate time investment.
  * Why? Security candidates are a difficult to source engineering disciplines. We want to reduce the chance they see our process and say "that's not worth the effort".
* Increase the signal we get about "how does this candidate review & communicate feedback?" 
  * Why? Given the nature of security roles, review and guidance will be a larger part of the job. Our traditional challenges are highly "greenfield" development based. We'd like to adjust this to gather data about candidates improving established code bases.

## Reference Resources

* https://goteleport.com/blog/coding-challenge/
* https://sockpuppet.org/blog/2015/03/06/the-hiring-post/

### CODE REVIEW

File: challenges/security/application-security.md [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+via internal promotion. See the [engineering levels document](https://raw.githubusercontent.com/gravitational/careers/main/levels.pdf)
+for more details.
+
+We expect different performance at different engineering levels. These expectations
+are explicitly documented below.
+
+
+# Part 1: Peer Review
+You will be given a GitHub repo containing a PR that includes a RFD and architecture
+diagram for a service similar to docker.
```
Comment (wadells) [inline challenges/security/application-security.md:48]: > What is this RFD and architecture diagram?

I haven't written it yet -- I'll be working on it today. The gist is a 2 page document targeting L3 of our [systems engineer challenge](https://github.com/gravitational/careers/blob/a78b45b4ddd5a513e66a437648b6d8600bb59cfd/challenges/systems/challenge.md).  Notably excluded will be authn/authz, transport security, and logging.

File: challenges/security/application-security.md [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+Review the service as written, assuming it will run on a single server.
+
+## Level 5
+The service will run in multiple geographies with unreliable network connections
+between them. Please address the security and business tradeoffs between
+authentication and authorization consistency architectures.
+
+
+# Part 2: Development
+You will be given a prototype implementation of the RFD reviewed in part 1.
```
Comment (wadells) [inline challenges/security/application-security.md:64]: > Can I see the prototype code?

What is this RFD and architecture diagram?

I haven't written it yet -- I'll be working on it today. Notably excluded will be authn/authz, transport security, and logging.

File: challenges/security/application-security.md [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+The prototype codebase will be provided in go and the final version should build
+and run on 64-bit Linux machines.
+
+# Timing
+
+You can split coding over a couple of weekdays or weekends and find time to ask
+questions and receive feedback.
+
+Once you join the Slack channel, you have between 1 week complete the challenge.
```
Comment (wadells) [inline challenges/security/application-security.md:242]:
```suggestion
Once you join the Slack channel, you have 1 week complete the challenge.
```

We're trying to keep this from creeping to a 2 week challenge.  Ideally, we'd have a it be a one day challenge, but we're not there yet.

File: challenges/security/cloud-security.md [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+Review the service as written, assuming it will run on a single server.
+
+## Level 5
+The service will run in multiple geographies with unreliable network connections
+between them. Please address the security and business trade-offs between
+authentication and authorization consistency architectures.
+
+
+# Part 2: Development
+You will be given a prototype implementation of the RFD reviewed in part 1.
```
Comment (reedloden) [inline challenges/security/cloud-security.md:63]: How do we make sure that this prototype doesn't lead to issues with people applying for the appsec role? Might be better to have a different part 1 entirely.

File: challenges/security/cloud-security.md [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+Authenticated users should be able to download any of these blobs, but not change
+any data.
+
+Configure your cloud provider to log artifact any artifact write and retain these
+logs for at least 90 days.
+
+## Level 5
+
+TBD.
```
[the remark this answers is not in the document]
Comment (reedloden) [reply challenges/security/cloud-security.md:87]: Perhaps cross-account logging/IAM? Or using other AWS services as well?

File: teams.md [hunk 6]
```diff
@@ -215,3 +190,18 @@ on:
   improving our security posture
   audit trail for generated and published assets and ensuring we're using
   encryption at rest in the right places.
+
+## Security
```
Comment (reedloden) [inline teams.md:194]: Is it worth talking about how this is different from Product Security team (I dislike the team name)? Or perhaps within Product Security, we talk about how there's a separate Security team?

File: challenges/security/cloud-security.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (reedloden) [reply challenges/security/cloud-security.md:87]: I think tying in other AWS services would be valuable here... Show the true in-depth knowledge of AWS. (or GCP equivalent)

File: challenges/security/cloud-security.md [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+We are not currently hiring L1-L3 security engineers and level 6 is only available
+via internal promotion. See the [engineering levels document](https://raw.githubusercontent.com/gravitational/careers/main/levels/security.pdf)
+for more details.
+
+We expect different performance at different engineering levels. These expectations
+are explicitly documented below.
+
+# Part 1: Peer Review
+You will be given a GitHub repo containing a PR that includes a RFD and architecture
```
[the remark this answers is not in the document]
Comment (wadells) [reply challenges/security/cloud-security.md:46]: No.  That was my plan originally, but we discussed switching the cloud security engineer to have a AWS based architecture with terraform.  I've been working on that Design and repo, and will update it here.

File: challenges/security/cloud-security.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (wadells) [reply challenges/security/cloud-security.md:87]: I like the idea of cross account IAM.  Increasing the number of AWS services they need to configure isn't feasible IMO with the single day challenge.  There isn't enough time to for candidates to hack up the half dozen I'd like to see:

* IAM
* S3
* EC2 / Security Groups / VPC
* Cloudtrail
* Guardduty
* Some sort of container based service - ECS/EKS/ECR/Fargate 

I think we can get 2, *maybe* 3 of these in 2-3 hours of development time.  If we opted for the asynchronous interview process, we could ask for more.

File: teams.md (hunk 6 above)
Comment (wadells) [reply teams.md:194]: Yes -- I'll cross link between the two sections.

Comment (wadells) [reply teams.md:197]: Cross linked and took a shot at explaining the differences in https://github.com/gravitational/careers/pull/66/commits/1f4cc986761af9e6db5017e5de31f1e468d76e0a

File: challenges/security/application-security.md [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+
+In case of a negative score result, hiring manager will contact you and share a
+list of the key observations from the team that affected the result.
+
+## Code and project ownership
+
+This is a test challenge and we have no intent of using the code you've
+submitted in production. This is your work, and you are free to do whatever you
+feel is reasonable with it. In the scenario where you don't pass, you can open
+source it with any license and use it as a portfolio project.
```
[the remark this answers is not in the document]
Comment (wadells) [reply challenges/security/application-security.md]: Updated to mention we retain rights to our sample code in https://github.com/gravitational/careers/pull/66/commits/bb7b1e0b02198ccfdf0d1d6d91e9dcac42818586

File: challenges/security/application-security.md (hunk 3 above)
Comment (wadells) [reply challenges/security/application-security.md]: This was rewritten to be a one day challenge in https://github.com/gravitational/careers/pull/66/commits/bb7b1e0b02198ccfdf0d1d6d91e9dcac42818586

File: challenges/security/cloud-security.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (wadells) [reply challenges/security/cloud-security.md]: I took a stab at fleshing out the different security features we'd like to see (while trying to remain AWS/GCP agnostic). See https://github.com/gravitational/careers/commit/ded4381259f0758e48a50281b3c3f5c71f6dd7cd

File: challenges/security/cloud-security.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (wadells) [reply challenges/security/cloud-security.md]: This has been changed in https://github.com/gravitational/careers/commit/ded4381259f0758e48a50281b3c3f5c71f6dd7cd

File: challenges/security/cloud-security.md [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+
+## Level 4
+
+A Make target should be able to store built artifacts using a least-privilege
+service account.
+
+Authenticated users should be able to download any of these blobs, but not change
+any data.
+
+Configure your cloud provider to log artifact any artifact write and retain these
```
[the remark this answers is not in the document]
Comment (wadells) [reply challenges/security/cloud-security.md]: Clarified audit logs in https://github.com/gravitational/careers/commit/ded4381259f0758e48a50281b3c3f5c71f6dd7cd

File: challenges/security/application-security.md [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+next section.
+
+# Part 4: Development
+You will be given a prototype go implementation of the RFD reviewed in
+part 1. You will have 3 hours to develop and submit a single PR implementing
+an authorization and authentication scheme.
+
+This is asynchronous. You'll develop on your machine and you're welcome to
+use whatever tools you're comfortable with to develop the changes. We'll be
+available in the interview slack channel to answer any questions.
```
Comment (adaadb6) [inline challenges/security/application-security.md:95]: Curious if the Slack channel should be mentioned at the top in the summary with a list of things to expect before the interviewing day?

File: challenges/security/cloud-security.md [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+## Areas of focus
+
+These are the areas we will be evaluating in the submission:
+
+* Ensure the Terraform contains all resources needed.
+* Use consistent coding style. We follow `terraform fmt` and
+  [tfsec](https://github.com/aquasecurity/tfsec) for our Terraform.
+* Follow the [principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege)
+  when configuring IAMs.
+* Show your breadth and depth of knowledge for either AWS or GCP.
```
[the remark this answers is not in the document]
Comment (wadells) [reply challenges/security/cloud-security.md:171]: This is tricky.

We definitely want candidates with AWS experience.  GCP experience is nice, but not as useful as AWS given our architecture.  I don't think we'd be in a good position to review or hire Azure folks, or any of the smaller providers.

So that is one point in favor of sticking with AWS, and maybe allowing GCP as a 2nd option.

Beyond the skills we need, I'm also considering the candidate experience and how many different processes we have to create and review.  Consider:

It takes a couple interviews to get well calibrated on when candidates are good and when they're not as strong as others.  We're already developing and allowing app sec vs cloud sec processes, which won't be directly comparable.  If we allow cloud sec to fragment into AWS and GCP, we're decreasing the chance the data we'll have to directly compare candidates, and introducing a chance for more subjectivity to creep into the process. AWS and GCP aren't 1-to-1, and how do we evaluate a candidate prioritizing GCP specific features vs AWS specific features.  I wouldn't want to say "You can use GCP" but then have it cause candidates to start off on a worse foot than those who chose AWS

The other element is the investment of developing architecture and example codebases.  It will take additional work to put together example code for a GCP version in addition to an AWS version.  Given that we're already months behind on starting to hire for these roles, I'm looking to cut scope rather than increase it.

We can avoid the cost of additional example code dev by having the terraform challenge be greenfield (no starting example code), but my gut says it is top much ask a candidate to do greenfield development in 3 hours.  If we went with the async challenge, I'd feel better about greenfield development, and thus allowing GCP.

@reedloden  and I  discussed allowing candidates to choose a synchronous or asyncrhonous process. However I'm reluctant to do that in this PR, as it would compound the difficulty of comparing candidates who choose different options.  Given that we're hiring 4 security engineers, we'd quickly reach the point where we could have only one or two candidates in any given process.

In short: Today, I'm developing the process only for AWS.  We may iterate on it to add GCP in the near future.

File: challenges/security/application-security.md (hunk 10 above)
Comment (wadells) [reply challenges/security/application-security.md:95]: Good suggestion -- I'll make this edit.

Comment (wadells) [reply challenges/security/application-security.md:95]: In the summary, there is the following stanza:

> Throughout the day, the interview team will join you in a slack channel, where we'll provide relevant links, and you're welcome to ask any questions.

Let me know if you'd adjust this wording at all.

File: challenges/security/application-security.md [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+We use a synchronous, remote, one day hiring challenge for application
+security engineers. The challenge consists of the following sections:
+
+1. Kick off with the hiring manager. 45 minutes. 09:15 - 10:00 PST
+
+2. Peer review. You'll receive a brief Request for Discussion (RFD) and
+architecture diagram of a service. You will have 90 minutes to
+provide written PR feedback. 10:00 - 11:30 PST.
+
+3. After the peer review, section, there is a 30 minute break for lunch.
```
Comment (reedloden) [inline challenges/security/application-security.md:12]:
```suggestion
3. After the peer review section, there is a 30 minute break for lunch.
```

File: challenges/security/application-security.md [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+
+4. Development. You'll receive prototype code for the service reviewed
+in part one. Building off this prototype codebase, we ask you develop
```
Comment (reedloden) [inline challenges/security/application-security.md:15]: "Part one" isn't defined. Should we say section 2? Or explicitly state it's different parts?

File: challenges/security/application-security.md [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+some changes and open a pull request over 3 hours. 12:00 - 15:00 PST
+
+5. Attack & defense discussion.  This is a 60 minute video call with future
```
Comment (reedloden) [inline challenges/security/application-security.md:18]:
```suggestion
5. Attack & defense discussion. This is a 60 minute video call with future
```

File: challenges/security/application-security.md [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+peers where we ask you to show your depth of knowledge in hypothetically
+attacking a service, and defending against those attacks.
+You will also have a chance to ask us questions about working at Teleport.
+15:00 - 16:00 PST.
```
Comment (reedloden) [inline challenges/security/application-security.md:22]: What is this time for?

File: challenges/security/application-security.md [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+
+Throughout the day, the interview team will join you in a slack channel, where
```
Comment (reedloden) [inline challenges/security/application-security.md:24]:
```suggestion
Throughout the day, the interview team will join you in a Slack channel, where
```

File: challenges/security/application-security.md [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+we'll provide relevant links, and you're welcome to ask any questions.
+
+# Rationale
+
+These exercises have several goals:
+
+* It helps us to understand what to expect from you as a developer, what
+  security issues you find important, how you communicate those issues to peers
```
Comment (reedloden) [inline challenges/security/application-security.md:32]:
```suggestion
  security issues you find important, how you communicate those issues to peers,
```

File: challenges/security/application-security.md [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+  and how you write production code.
+* It helps you get a feel for what it would be like to work at Teleport, as this
+  exercise aims to simulate our day-as-usual and expose you to the type of work
```
Comment (reedloden) [inline challenges/security/application-security.md:35]:
```suggestion
  exercise aims to simulate our day-as-usual and exposes you to the type of work
```

File: challenges/security/application-security.md [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+between them. Please address the security and business tradeoffs between
+authentication and authorization consistency architectures.
+
+# Part 3: Break
+Grab a bite of food and some water. Take a breather before we continue to the
+next section.
+
+# Part 4: Development
+You will be given a prototype go implementation of the RFD reviewed in
+part 1. You will have 3 hours to develop and submit a single PR implementing
```
Comment (reedloden) [inline challenges/security/application-security.md:90]: "part 1" is the hiring manager kick off.

File: challenges/security/application-security.md [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+an authorization and authentication scheme.
+
+This is asynchronous. You'll develop on your machine and you're welcome to
```
Comment (reedloden) [inline challenges/security/application-security.md:93]:
```suggestion
This is asynchronous. You'll develop on your machine, and you're welcome to
```

File: challenges/security/application-security.md (hunk 10 above)
Comment (reedloden) [inline challenges/security/application-security.md:95]:
```suggestion
available in the interview Slack channel to answer any questions.
```

File: challenges/security/application-security.md [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+use whatever tools you're comfortable with to develop the changes. We'll be
+available in the interview slack channel to answer any questions.
+
+We will review your PR, including compiling and run the program and testing
+it behavior. Make sure to write good commit messages, as well as a helpful
+PR description.
+
+This prototype is expected to run on only a single machine. The distributed
+state considerations of Part 2 L5 need not be considered.
```
Comment (reedloden) [inline challenges/security/application-security.md:102]: Clarify what L5 is here.

File: challenges/security/cloud-security.md [hunk 22]
```diff
@@ -0,0 +1,241 @@
+# Summary
```
Comment (reedloden) [inline challenges/security/cloud-security.md:1]: Similar issues as above. Not repeating them.

File: challenges/security/application-security.md [hunk 23]
```diff
@@ -0,0 +1,270 @@
+# Summary
```
Comment (reedloden) [inline challenges/security/application-security.md:1]: Not sure what happened, but all my edits got lost somehow. I had made a lot of spelling and grammar fixes in a previous commit. I'll download this and do a pass separately.

File: challenges/security/cloud-security.md [hunk 24]
```diff
@@ -0,0 +1,241 @@
...
+
+# Part 1: Hiring Manager Kick Off
+The hiring manager will discuss your background and approach to security with
+you.
+
+
+# Part 2: Peer Review
+You will be given a GitHub repo containing a PR that includes a
+[RFD](https://github.com/gravitational/teleport/blob/master/rfd/0000-rfds.md)
+and architecture diagram for a cloud hosted binary distribution service.
```
Comment (reedloden) [inline challenges/security/cloud-security.md:68]:
```suggestion
and architecture diagram for a cloud-hosted binary distribution service.
```

File: challenges/security/cloud-security.md [hunk 25]
```diff
@@ -0,0 +1,241 @@
...
+
+# Part 4: Development
+You will be given a prototype AWS terraform implementation of the RFD reviewed in
+part 1. You will have 3 hours to develop and submit a single PR securing the
+service. Please implement as many of the following security features as you
+find time for. Implement the ones that you believe provide the best security
+return on investment.
+
+ - Tamper-resitant audit logs
+ - Principle of least privilege, role based IAMS for:
```
Comment (reedloden) [inline challenges/security/cloud-security.md:97]:
```suggestion
 - Principle of least privilege, role-based IAM for:
```

File: challenges/security/cloud-security.md [hunk 26]
```diff
@@ -0,0 +1,241 @@
...
+   - uploading new artifacts
+   - reading artifacts
+   - viewing audit logs
+ - Encryption at Rest
+ - Encryption in Transit
+ - [Write Once Read Many](https://en.wikipedia.org/wiki/Write_once_read_many)
+   or similar data retention protections to avoid data loss
+ - Cloud provider specific security services such as:
+   - AWS: CloudTrail, GuardDuty, WAF
```
Comment (reedloden) [inline challenges/security/cloud-security.md:106]:
```suggestion
   - AWS: CloudTrail, Config, GuardDuty, WAF
```

File: challenges/security/cloud-security.md [hunk 27]
```diff
@@ -0,0 +1,241 @@
...
+
+This is asynchronous. You'll develop on your machine and you're welcome to
+use whatever tools you're comfortable with to develop the changes. We'll be
+available in the interview slack channel to answer any questions.
+
+We will review your PR, inculding running terraform plan. Make sure to write
```
Comment (reedloden) [inline challenges/security/cloud-security.md:112]:
```suggestion
We will review your PR, including running `terraform plan`. Make sure to write

File: challenges/security/application-security.md (hunk 23 above)
Comment (wadells) [reply challenges/security/application-security.md:1]: Sorry -- I must have force pushed over the changes when I rebased  at https://github.com/gravitational/careers/pull/66#event-7224472835 -- I typically try to pick them up.

I think I can probably fish them out of https://github.com/gravitational/careers/commit/952d8e7f53aa7c973a1ccd6488b0ae849ca62c70

Comment (wadells) [reply challenges/security/application-security.md:1]: Reintroduced in https://github.com/gravitational/careers/pull/66/commits/66bf056ee02af1554320e364a53269a4b3ea685e

File: challenges/security/cloud-security.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (reedloden) [reply challenges/security/cloud-security.md:171]: :+1: to just focusing on AWS for now.

File: challenges/security/application-security.md (hunk 15 above)
Comment (wadells) [reply challenges/security/application-security.md]: This is when the attack and defense discussion will occur. Just like all the other sections have a time defined, so does this final step.

File: challenges/security/application-security.md [hunk 28]
```diff
@@ -0,0 +1,270 @@
...
+
+We use a synchronous, remote, one day hiring challenge for application
+security engineers. The challenge consists of the following sections:
+
+1. Kick off with the hiring manager. 45 minutes. 09:15 - 10:00 PST
```
[the remark this answers is not in the document]
Comment (wadells) [reply challenges/security/application-security.md]: Changed to PT in https://github.com/gravitational/careers/pull/66/commits/d1610474aab4c04d9c5f2623545583439fecb394.

File: challenges/security/application-security.md (hunk 19 above)
Comment (wadells) [reply challenges/security/application-security.md]: I made the language position agnostic.  Now is is "the RFD you reviewed earlier".

File: challenges/security/application-security.md (hunk 21 above)
Comment (wadells) [reply challenges/security/application-security.md]: Changed to "Level 5". Let me know if further clarification is needed.

File: challenges/security/application-security.md (hunk 13 above)
Comment (wadells) [reply challenges/security/application-security.md]: Changed to "the service you reviewed earlier".

**Review by reedloden — APPROVED** [review]
